employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/NullableValueElement.java

```diff
@@ -0,0 +1,118 @@
+package org.optaweb.employeerostering.gwtui.client.common;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+import javax.inject.Inject;
+
+import com.google.gwt.user.client.TakesValue;
+import elemental2.dom.Event;
+import elemental2.dom.HTMLInputElement;
+import elemental2.dom.HTMLLabelElement;
+import org.jboss.errai.ui.shared.api.annotations.DataField;
+import org.jboss.errai.ui.shared.api.annotations.EventHandler;
+import org.jboss.errai.ui.shared.api.annotations.ForEvent;
+import org.jboss.errai.ui.shared.api.annotations.Templated;
+
+@Templated
+public class NullableValueElement<T> implements TakesValue<Optional<T>> {
+
+    @Inject
+    @DataField("has-value-checkbox")
+    private HTMLInputElement hasValueCheckbox;
+
+    @Inject
+    @DataField("value")
+    private HTMLInputElement valueInput;
+
+    @Inject
+    @DataField("label")
+    private HTMLLabelElement label;
+
+    private Optional<T> value;
+
+    private Function<String, T> valueConverter;
+
+    @EventHandler("has-value-checkbox")
+    public void onHasValueCheckboxUpdate(@ForEvent("change") Event e) {
+        valueInput.disabled = !hasValueCheckbox.checked;
+        if (valueInput.disabled) {
+            value = Optional.empty();
+            valueInput.required = false;
+        } else {
+            value = getValueFromInput();
+            valueInput.required = true;
+        }
+    }
+
+    @EventHandler("value")
+    public void onValueChange(@ForEvent("change") Event e) {
+        value = getValueFromInput();
+    }
+
+    public void init(Class<T> clazz, T value, String labelText) {
```

Comment:
Apart from using string comparison instead of `isAssignableFrom()` method, I wonder if we need such a complicated init method at all. Generally, conditions base on types can be replaced by class hierarchy - in this case for Integer, Double and String.

Reply:
GWT does not support the `isAssignableFrom()` method unfortunately. An unfortunate side effect of using class hierarchy is that the HTML template file would need to be duplicated with no changes and each would need to be maintained separately (although currently we only actually use the Integer version).

Reply:
Does it mean this change is not possible?
```diff
    @Inject
    @DataField("maximum-minutes-per-day")
-    private NullableValueElement<Integer> maximumMinutesPerDay;
    @Inject
    @DataField("maximum-minutes-per-day")
+    private NullableIntegerElement maximumMinutesPerDay;
```
Would it affect the HTML template? Why would I need to duplicate it?

Reply:
Nevermind; `@Templated` can take a fully qualified path, which would solve the duplication issue. (the issue have to do with the implementation of Nullable<Type>Element, which would be a Templated bean). Actually can probably remove the html code duplication from all the Entity List pages using this.

Reply:
Thanks, I have a better understanding now. Would it be possible to make the common ancestor (an abstract class `NullableElement`) `@Templated`, and still use its child, `NullableIntegerElement` like this:
```java
    @Inject
    @DataField("maximum-minutes-per-day")
    private NullableIntegerElement maximumMinutesPerDay;
```
even if `NullableIntegerElement` is not `@Templated`? If that worked, it would allow you to avoid using `@Templated("f.q.c.N")` on each concrete implementation of `NullableElement`.

Reply:
`@Templated("f.q.c.N")` is not the problem, but the solution (it prevent duplication of HTML files). Classes do not inherit annotations, so it is impossible to make the children inherit the `@Templated` annotation. Instead, the parent remains unannotated and we annotate the children.

Reply:
Makes sense. I know that annotations are not inherited but Errai could have been smart and traverse the class hierarchy towards root if I inject an unannotated class as a field in a `@Templated` class.

Your proposal seems good to me though.

Reply:
@Christopher-Chianelli if there is no way to replace the class name comparison by a class hierarchy due to limitations of the GWT and Errai framework, would it be at least possible to replace the `getSimpleName()` with an `instanceof` operator? In contrast to the `isAssignableFrom()`, for using `instanceof` the class must be known at compile time, so maybe GWT might support it.